README
# RarCrack!

This program licensed under GPL2+.

If you forget your password for compressed archive (rar, 7z, zip), this program is the solution.
This program uses bruteforce algorithm to find correct password. You can specify wich characters will be used in password generations.
Warning: Please don't use this program for any illegal things!

## Changelog

See CHANGELOG.md

## Software requirements
* C99 compiler or C11 compiler (gcc is recommanded)
* > glibc 2.4
* any POSIX compatible operating system [sorry Window$ isn't]
* pthreads
* libxml2
* and finally: 7zip, unrar, unzip

## Building and installing

Everything is very easy:
---------------------------------------------------
tar -xjf rarcrack-VERSION.tar.bz2
cd rarcrack-VERSION
#you need gcc or any C compiler (edit Makefile CC=YOUR_C_COMPILER)
make
#you must be root in next step:
make install
---------------------------------------------------


## Using
---------------------------------------------------
rarcrack your_encrypted_archive.ext [--threads thread_num] [--type rar|zip|7z]
---------------------------------------------------
Everything in [] are optional, rarcrack default crack two threads and autodetect the archive type. If the detection wrong you can specify the correct file type with the type parameter. RarCrack currently crack maximum in 12 threads.

After the cracking started RarCrack will print the current status of cracking and save it's to a status file. If you want more specific password character set, you need to run RarCrack to create the XML status file (3 sec).
---------------------------------------------------
<?xml version="1.0" encoding="UTF-8"?>
<rarcrack>
 <abc>0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ</abc>
 <current>uU</current>
 <good_password></good_password>
</rarcrack>
---------------------------------------------------
This is a sample XML file, and you see there is a character set. If you want, you can modify this file and when you start RarCrack again the program will be use new variables.
Warning: Take care when you changing this file, make sure the current password don't have characters outside the abc[character set]!


CHANGELOG.md
# Changelog

## Versions

* Version 0.2.1
  * Uncompatible with C89, but compatible with C99 and C11
  * Minor modifications
* Version 0.2, by David Zoltan Kedves
  * Added multiple cracking threads support
  * Added 7z, zip support
  * Archive file type auto-detection
* Version 0.1, by David Zoltan Kedves
  * Inital version (rar cracking working)

## Authors

* David Zoltan Kedves (kedazo@gmail.com)
* Spanti Nicola (rydroid_dev@yahoo.com)


rarcrack.h
/**
 * Copyright (C) 2007 by David Zoltan Kedves
 * kedazo@gmail.com
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the
 * Free Software Foundation, Inc.,
 * 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#define APP_NAME    "RarCrack!"
#define APP_VERSION "0.2.1"
#define APP_LICENSE "GPL2+"

#define _BSD_SOURCE // popen with C99 and glibc
#define _GNU_SOURCE // strcasestr with C99 and glibc

#include <stdio.h> // Standard headers
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <libxml/xmlmemory.h> // libxml2 headers
#include <libxml/parser.h>
#include <libxml/parserInternals.h>
#include <libxml/tree.h>
#include <libxml/threads.h>
#include <pthread.h> // POSIX threads

char default_ABC[] = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

const char CMD_DETECT[] = "file -i -b %s"; //this command return what is the file mime type

const char* TYPE[] = {"rar",		"7z",		"zip", ""}; //the last "" signing this is end of the list
const char* MIME[] = {"application/x-rar", "application/octet-stream", "application/x-zip", ""};
const char* CMD[] = {"unrar t -y -p%s %s 2>&1", "7z t -y -p%s %s 2>&1", "unzip -P%s -t %s 2>&1", ""};

#define PWD_LEN 100


rarcrack.c
/**
 * Copyright (C) 2007 by David Zoltan Kedves
 * kedazo@gmail.com
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the
 * Free Software Foundation, Inc.,
 * 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include "rarcrack.h"

char* ABC = (char*) &default_ABC;
int ABCLEN;

char password[PWD_LEN+1] = {'\0','\0'}; //this contains the actual password
char password_good[PWD_LEN+1] = {'\0', '\0'};  //this changed only once, when we found the good passord
unsigned int curr_len = 1; //current password length
long counter = 0; //this couning probed passwords
xmlMutexPtr pwdMutex; //mutex for password char array
char filename[255];	//the archive file name
char statname[259];	//status xml file name filename + ".xml"
xmlDocPtr status;
int finished = 0;
xmlMutexPtr finishedMutex;
char finalcmd[300] = {'\0', '\0'}; //this depending on arhive file type, it's a command to test file with password

char * getfirstpassword() {
	static char ret[2];
	ret[0] = ABC[0];
	ret[1] = '\0';
	return (char*) &ret;
}

inline void savestatus() {
	if ((strlen(statname) > 0) && (status != NULL)) {
		xmlNodePtr root = xmlDocGetRootElement(status);
		if (root) {
			xmlMutexLock(finishedMutex);
			xmlChar* tmp = NULL;
			for (xmlNodePtr node = root->children; node; node = node->next) {
				if (xmlStrcmp(node->name, "current") == 0) {
					xmlMutexLock(pwdMutex);
					tmp = xmlEncodeEntitiesReentrant(status, (const xmlChar*) &password);
					xmlMutexUnlock(pwdMutex);
					if (node->children) {
						if (password[0] == '\0')
							xmlNodeSetContent(node->children, getfirstpassword());
						else
							xmlNodeSetContent(node->children, tmp);
						}
					xmlFree(tmp);
				} else if ((finished == 1) && (xmlStrcmp(node->name,"good_password") == 0)) {
					tmp =  xmlEncodeEntitiesReentrant(status, (const xmlChar*) &password_good);
					if (node->children)
						xmlNodeSetContent(node->children, tmp);
					xmlFree(tmp);
				}
			}
			xmlMutexUnlock(finishedMutex);
		}
		xmlSaveFormatFileEnc(statname, status, "UTF-8", 1);
	}
}

inline unsigned int abcnumb(char a) {
	for (unsigned int i = 0; i < ABCLEN; ++i)
	{
		if (ABC[i] == a)
			return i;
	}
	return 0;
}

int loadstatus() {
	xmlNodePtr root = NULL;
	xmlNodePtr node = NULL;
	
	int ret = 0;
	FILE* totest = fopen(statname, "r");
	if (totest != NULL) {
		fclose(totest);
		status = xmlParseFile(statname);
	}
	if (status != NULL)
		root = xmlDocGetRootElement(status);
	else
		status = xmlNewDoc(NULL);
	if (root != NULL) {
		char* tmp;
		xmlParserCtxtPtr parserctxt = xmlNewParserCtxt();
		for (node = root->children; node; node = node->next) {
			if (xmlStrcmp(node->name, "abc") == 0) {
				if (node->children && (strlen(node->children->content) > 0)) {
					ABC = xmlStringDecodeEntities(
						parserctxt,
						node->children->content,XML_SUBSTITUTE_BOTH,
						0,0,0);
				} else
					ret = 1;
			} else if (xmlStrcmp(node->name, "current") == 0) {
				if (node->children && (strlen(node->children->content) > 0)) {
					tmp = xmlStringDecodeEntities(
						parserctxt,
						node->children->content,XML_SUBSTITUTE_BOTH,
						0,0,0);
					strcpy(password,tmp);
					curr_len = strlen(password);
					printf("INFO: Resuming cracking from password: '%s'\n", password);
					xmlFree(tmp);
				} else
					ret = 1;
			} else if (xmlStrcmp(node->name, "good_password") == 0) {
				if (node->children && (strlen(node->children->content) > 0)) {
					tmp = xmlStringDecodeEntities(
						parserctxt,
						node->children->content,XML_SUBSTITUTE_BOTH,
						0,0,0);
					strcpy(password, tmp);
					curr_len = strlen(password);
					xmlMutexLock(finishedMutex);
					finished = 1;
					xmlMutexUnlock(finishedMutex);
					strcpy((char*) &password_good, (char*) &password);
					printf("GOOD: This archive was succesfully cracked\n");
					printf("      The good password is: '%s'\n", password);
					xmlFree(tmp);
					ret = 1;
				}
			}
		}
		xmlFreeParserCtxt(parserctxt);
	} else {
		root = xmlNewNode(NULL, "rarcrack");
		xmlDocSetRootElement(status, root);
		node = xmlNewTextChild(root, NULL, "abc", ABC);
		node = xmlNewTextChild(root, NULL, "current", getfirstpassword()); 
		node = xmlNewTextChild(root, NULL, "good_password", "");
		savestatus();
	}
	
	return ret;
}

void nextpass2(char* p, unsigned int n) {
	if (p[n] == ABC[ABCLEN-1]) {
		p[n] = ABC[0];
		if (n > 0)
			nextpass2(p, n-1);
		else {
			for (int i=curr_len; i >= 0; --i)
				p[i+1] = p[i];
				p[0] = ABC[0];
				p[++curr_len] = '\0';
		}
	}
	else
	{
		p[n] = ABC[abcnumb(p[n])+1];
	}
}

inline char* nextpass() {	//IMPORTANT: the returned string must be freed
	char *ok = malloc(sizeof(char)*(PWD_LEN+1));
	xmlMutexLock(pwdMutex);
	strcpy(ok, password);
	nextpass2((char*) &password, curr_len - 1);
	xmlMutexUnlock(pwdMutex);
	return ok;
}

void * status_thread() {
	const short status_sleep = 3;
	
	while(1) {
		sleep(status_sleep);
		xmlMutexLock(finishedMutex);
		if (finished != 0)
			break;
		xmlMutexUnlock(finishedMutex);
		xmlMutexLock(pwdMutex);
		printf("Probing: '%s' [%d passwords/sec]\n", password, counter / status_sleep);
		counter = 0;
		xmlMutexUnlock(pwdMutex);
		savestatus();	//FIXME: this is wrong, when probing current password(s) is(are) not finished yet, and the program is exiting
	}
}

void * crack_thread() {
	char * current;
	char ret[200];
	char cmd[400];
	FILE * Pipe;
	
	while (1) {
		current = nextpass();
		sprintf((char*)&cmd, finalcmd, current, filename);
		Pipe = popen(cmd, "r");
		while (!feof(Pipe)) {
			fgets((char*)&ret, 200, Pipe);
			if (strcasestr(ret, "ok") != 0) { // NULL
				strcpy(password_good, current);
				xmlMutexLock(finishedMutex);
				finished = 1;
				printf("GOOD: password cracked: '%s'\n", current);
				xmlMutexUnlock(finishedMutex);
				savestatus();
				break;
			}
		}
		pclose(Pipe);
		xmlMutexLock(finishedMutex);
		++counter;
		if (finished != 0) {
			xmlMutexUnlock(finishedMutex);
			break;
		}
		xmlMutexUnlock(finishedMutex);
		free(current);
	}
}

void crack_start(unsigned int threads) {
	pthread_t th[13];
	unsigned int i;
	for (i = 0; i < threads; ++i) {
		(void) pthread_create(&th[i], NULL, crack_thread, NULL);
	}
	(void) pthread_create(&th[12], NULL, status_thread, NULL);
	for (i = 0; i < threads; ++i) {
		(void) pthread_join(th[i], NULL);
	}
	(void) pthread_join(th[12], NULL);
}

void print_option(char* short_option, char* long_option, char* arg, char* text)
{
	if((short_option != NULL && strlen(short_option) > 0) || (long_option != NULL && strlen(long_option) > 0)) {
		printf("  ");
		if(short_option != NULL && strlen(short_option) > 0)
			printf("-%s", short_option);
		if(short_option != NULL && strlen(short_option) > 0 && long_option != NULL && strlen(long_option) > 0)
			printf(" ");
		if(long_option != NULL && strlen(long_option) > 0)
			printf("--%s", long_option);
		if(arg != NULL && strlen(arg) > 0)
			printf("  %s", arg);
		printf("\n");
		if(text != NULL && strlen(text) > 0)
			printf("    %s\n", text);
	}
}

void init(int argc, char **argv) {
	xmlInitThreads();
	pwdMutex = xmlNewMutex();
	finishedMutex = xmlNewMutex();
	
	unsigned int i, j;
	int help = 0;
	int threads = 4;
	int archive_type = -1;
	char test[500];
	
	if (argc == 1) {
		printf("USAGE: rarcrack encrypted_archive.ext [--threads NUM] [--type rar|zip|7z]\n");
		printf("\tFor more information please run \"rarcrack --help\"\n");
		help = 1;
	} else {
		for (i = 1; i < argc; ++i) {
			if (strcmp(argv[i], "--help") == 0 || strcmp(argv[i], "-h") == 0) {
				printf("Usage:\n");
				printf("  rarcrack encrypted_archive.ext [--threads NUM] [--type rar|zip|7z]\n\n");
				printf("Options:\n");
				print_option("h", "help", "",  "show this screen");
				print_option("",  "type", "zip|rar|7z", "you can specify the archive program, this needed when the program couldn't detect the proper file type");
				print_option("",  "threads", "num>0", "you can specify how many threads will be run (maximum: 12, default: 4)");
				print_option("v", "version", "",  "version of this program");
				print_option("",  "license", "",  "license of this program");
				printf("\nInfo: This program supports only RAR, ZIP and 7Z encrypted archives.\n");
				printf("%s usually detects the archive type.\n\n", APP_NAME);
				help = 1;
				break;
			} else if (strcmp(argv[i], "--version") == 0 || strcmp(argv[i],"-v") == 0) {
				printf("%s, version %s\n", APP_NAME, APP_VERSION);
				help = 1;
				break;
			} else if (strcmp(argv[i], "--license") == 0) {
				printf("%s is under GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.\n", APP_NAME);
				help = 1;
				break;
			} else if (strcmp(argv[i], "--threads") == 0) {
				if ((i + 1) < argc) {
					sscanf(argv[++i], "%d", &threads);
					if (threads < 1) {
						printf("INFO: number of threads adjusted to 1\n");
						threads = 1;
					} else if (threads > 12) {
						printf("INFO: number of threads adjusted to 12\n");
						threads = 12;
					}
				} else {
					printf("ERROR: missing parameter for option: --threads!\n");
					help = 1;
				}
			} else if (strcmp(argv[i],"--type") == 0) {
				if ((i + 1) < argc) {
					sscanf(argv[++i], "%s", &test);
					for (j = 0; strcmp(TYPE[j], "") != 0; ++j) {
						if (strcmp(TYPE[j], test) == 0) {
							strcpy(finalcmd, CMD[j]);
							archive_type = j;
							break;
						}
					}
					if (archive_type < 0) {
						printf("WARNING: invalid parameter --type %s!\n", argv[i]);
						finalcmd[0] = '\0';
					}
				} else {
					printf("ERROR: missing parameter for option: --type!\n");
					help = 1;
				}
			} else {
				strcpy((char*)&filename, argv[i]);
			}
		}
	}
	if (help == 1)
		return;
	sprintf((char*)&statname,"%s.xml", (char*)&filename);
	FILE* totest = fopen(filename, "r");
	if (totest == NULL) {
		printf("ERROR: The specified file (%s) is not exists or \n", filename);
		printf("\tyou don't have the right permissions!\n");
		return;
	} else
		fclose(totest);
	if (finalcmd[0] == '\0') { // when we specify the file type, the programm will skip the test
		sprintf((char*)&test, CMD_DETECT, filename);
		totest = popen(test, "r");
		fscanf(totest,"%s",(char*)&test);
		pclose(totest);
		for (i = 0; strcmp(MIME[i],"") != 0; ++i) {
			if (strcmp(MIME[i], test) == 0) {
				strcpy(finalcmd, CMD[i]);
				archive_type = i;
				break;
			}
		}
		printf("INFO: detected file type: %s\n", TYPE[archive_type]);
	} else
		printf("INFO: the specified archive type: %s\n", TYPE[archive_type]);
	if (finalcmd[0] == '\0') {
		printf("ERROR: Couldn't detect archive type\n");
		return;
	} /*else
		printf("DEBUG: the unpack command is: '%s'\n", finalcmd);*/
	printf("INFO: cracking %s, status file: %s\n", filename, statname);
	if (loadstatus() == 1) {
		printf("ERROR: The status file (%s) is corrupted!\n", statname);
		return;
	}
	ABCLEN = strlen(ABC);
	if (password[0] == '\0')
		password[0] = ABC[0];
	crack_start(threads);
}

int main(int argc, char *argv[]) {
	printf("%s %s (rarcrack.sourceforge.net) (%s)\n", APP_NAME, APP_VERSION, APP_LICENSE);
	printf("By David Zoltan Kedves (kedazo@gmail.com)\n\n");
	
	init(argc,argv);
	
	if (ABC != (char*) &default_ABC)
	  xmlFree(ABC);
	if (status)
	  xmlFreeDoc(status);
	xmlFreeMutex(pwdMutex);
	xmlFreeMutex(finishedMutex);
	
	return EXIT_SUCCESS;
}
